File: gdext/src/RdRenderer.cpp
#include "RdRenderer.h"
#include "common.h"
#include <array>
#include <godot_cpp/classes/object.hpp>
#include <godot_cpp/classes/rd_attachment_format.hpp>
#include <godot_cpp/classes/rd_pipeline_color_blend_state.hpp>
#include <godot_cpp/classes/rd_pipeline_color_blend_state_attachment.hpp>
#include <godot_cpp/classes/rd_pipeline_depth_stencil_state.hpp>
#include <godot_cpp/classes/rd_pipeline_multisample_state.hpp>
#include <godot_cpp/classes/rd_pipeline_rasterization_state.hpp>
#include <godot_cpp/classes/rd_sampler_state.hpp>
#include <godot_cpp/classes/rd_shader_file.hpp>
#include <godot_cpp/classes/rd_shader_spirv.hpp>
#include <godot_cpp/classes/rd_uniform.hpp>
#include <godot_cpp/classes/rd_vertex_attribute.hpp>
#include <godot_cpp/classes/rendering_device.hpp>
#include <godot_cpp/classes/rendering_server.hpp>
#include <godot_cpp/classes/resource_loader.hpp>
#include <godot_cpp/variant/utility_functions.hpp>
#include <imgui.h>
#include <unordered_map>
#include <unordered_set>

using namespace godot;

namespace ImGui::Godot {

struct RdRenderer::Impl
{
    RID shader;
    int64_t vtxFormat = 0;
    RID pipeline;
    RID sampler;
    TypedArray<RID> srcBuffers;
    TypedArray<RDUniform> uniforms;
    godot::PackedColorArray clearColors;

    std::unordered_map<RID, RID> framebuffers;
    std::unordered_map<ImTextureID, RID> uniformSets;
    std::unordered_set<ImTextureID> usedTextures;

    godot::PackedInt64Array vtxOffsets;

    RID idxBuffer;
    int idxBufferSize = 0; // size in indices
    RID vtxBuffer;
    int vtxBufferSize = 0; // size in vertices

    void SetupBuffers(ImDrawData* drawData);

    Impl()
    {
        clearColors.push_back(godot::Color(0, 0, 0, 0));
        srcBuffers.resize(3);
        vtxOffsets.resize(3);
    }
};

RID RdRenderer::GetFramebuffer(RID vprid)
{
    if (!vprid.is_valid())
        return RID();

    const RenderingServer* RS = RenderingServer::get_singleton();
    RenderingDevice* RD = RS->get_rendering_device();
    auto it = impl->framebuffers.find(vprid);
    if (it != impl->framebuffers.end())
    {
        RID fb = it->second;
        if (RD->framebuffer_is_valid(fb))
            return fb;
    }

    const RID vptex = RS->texture_get_rd_texture(RS->viewport_get_texture(vprid));
    godot::TypedArray<godot::RID> arr;
    arr.push_back(vptex);
    RID fb = RD->framebuffer_create(arr);
    impl->framebuffers[vprid] = fb;
    return fb;
}

void RdRenderer::Impl::SetupBuffers(ImDrawData* drawData)
{
    RenderingDevice* RD = RenderingServer::get_singleton()->get_rendering_device();

    int globalIdxOffset = 0;
    int globalVtxOffset = 0;

    const int idxBufSize = drawData->TotalIdxCount * sizeof(ImDrawIdx);
    godot::PackedByteArray idxBuf;
    idxBuf.resize(idxBufSize);

    const int vertBufSize = drawData->TotalVtxCount * sizeof(ImDrawVert);
    godot::PackedByteArray vertBuf;
    vertBuf.resize(vertBufSize);

    for (int i = 0; i < drawData->CmdListsCount; ++i)
    {
        ImDrawList* cmdList = drawData->CmdLists[i];

        std::copy(cmdList->VtxBuffer.begin(),
                  cmdList->VtxBuffer.end(),
                  reinterpret_cast<ImDrawVert*>(vertBuf.ptrw() + globalVtxOffset));
        globalVtxOffset += cmdList->VtxBuffer.size_in_bytes();

        std::copy(cmdList->IdxBuffer.begin(),
                  cmdList->IdxBuffer.end(),
                  reinterpret_cast<ImDrawIdx*>(idxBuf.ptrw() + globalIdxOffset));
        globalIdxOffset += cmdList->IdxBuffer.size_in_bytes();

        // create a uniform set for each texture
        for (int cmdi = 0; cmdi < cmdList->CmdBuffer.Size; ++cmdi)
        {
            const ImDrawCmd& drawCmd = cmdList->CmdBuffer[cmdi];
            const ImTextureID texid = drawCmd.GetTexID();
            if (!texid)
                continue;
            const RID texrid = make_rid(texid);
            if (!RD->texture_is_valid(texrid))
                continue;

            usedTextures.insert(texid);
            if (!uniformSets.contains(texid))
            {
                Ref<RDUniform> uniform;
                uniform.instantiate();
                uniform->set_binding(0);
                uniform->set_uniform_type(RenderingDevice::UNIFORM_TYPE_SAMPLER_WITH_TEXTURE);
                uniform->add_id(sampler);
                uniform->add_id(texrid);

                uniforms[0] = uniform;
                uniformSets[texid] = RD->uniform_set_create(uniforms, shader, 0);
            }
        }
    }

    RD->buffer_update(idxBuffer, 0, idxBuf.size(), idxBuf);
    RD->buffer_update(vtxBuffer, 0, vertBuf.size(), vertBuf);
}

RdRenderer::RdRenderer() : impl(std::make_unique<Impl>())
{
}

bool RdRenderer::Init()
{
    RenderingDevice* RD = RenderingServer::get_singleton()->get_rendering_device();
    if (!RD)
        return false;

    // set up everything to match the official Vulkan backend as closely as possible

    Ref<RDShaderFile> shaderFile =
        ResourceLoader::get_singleton()->load("res://addons/imgui-godot/data/ImGuiShader.glsl");

    impl->shader = RD->shader_create_from_spirv(shaderFile->get_spirv());

    if (!impl->shader.is_valid())
        return false;

    TypedArray<RDVertexAttribute> vattrs;

    RDVertexAttribute attr_points;
    attr_points.set_location(0);
    attr_points.set_format(RenderingDevice::DATA_FORMAT_R32G32_SFLOAT);
    attr_points.set_stride(sizeof(ImDrawVert));
    attr_points.set_offset(0);

    RDVertexAttribute attr_uvs;
    attr_uvs.set_location(1);
    attr_uvs.set_format(RenderingDevice::DATA_FORMAT_R32G32_SFLOAT);
    attr_uvs.set_stride(sizeof(ImDrawVert));
    attr_uvs.set_offset(sizeof(float) * 2);

    RDVertexAttribute attr_colors;
    attr_colors.set_location(2);
    attr_colors.set_format(RenderingDevice::DATA_FORMAT_R8G8B8A8_UNORM);
    attr_colors.set_stride(sizeof(ImDrawVert));
    attr_colors.set_offset(sizeof(float) * 4);

    vattrs.append(&attr_points);
    vattrs.append(&attr_uvs);
    vattrs.append(&attr_colors);

    impl->vtxFormat = RD->vertex_format_create(vattrs);

    RDPipelineColorBlendStateAttachment bsa;
    bsa.set_enable_blend(true);
    bsa.set_src_color_blend_factor(RenderingDevice::BLEND_FACTOR_SRC_ALPHA);
    bsa.set_dst_color_blend_factor(RenderingDevice::BLEND_FACTOR_ONE_MINUS_SRC_ALPHA);
    bsa.set_color_blend_op(RenderingDevice::BLEND_OP_ADD);
    bsa.set_src_alpha_blend_factor(RenderingDevice::BLEND_FACTOR_ONE);
    bsa.set_dst_alpha_blend_factor(RenderingDevice::BLEND_FACTOR_ONE_MINUS_SRC_ALPHA);
    bsa.set_alpha_blend_op(RenderingDevice::BLEND_OP_ADD);

    Ref<RDPipelineColorBlendState> blend;
    blend.instantiate();
    blend->set_blend_constant(Color(0.0f, 0.0f, 0.0f, 0.0f));
    blend->get_attachments().append(&bsa);

    Ref<RDPipelineRasterizationState> raster_state;
    raster_state.instantiate();
    raster_state->set_front_face(RenderingDevice::POLYGON_FRONT_FACE_COUNTER_CLOCKWISE);

    Ref<RDAttachmentFormat> af;
    af.instantiate();
    af->set_format(RenderingDevice::DATA_FORMAT_R8G8B8A8_UNORM);
    af->set_samples(RenderingDevice::TEXTURE_SAMPLES_1);
    af->set_usage_flags(RenderingDevice::TEXTURE_USAGE_COLOR_ATTACHMENT_BIT);

    TypedArray<RDAttachmentFormat> afs;
    afs.push_back(af);

    const int64_t fb_format = RD->framebuffer_format_create(afs);

    impl->pipeline = RD->render_pipeline_create(impl->shader,
                                                fb_format,
                                                impl->vtxFormat,
                                                RenderingDevice::RENDER_PRIMITIVE_TRIANGLES,
                                                raster_state,
                                                {},
                                                {},
                                                blend);

    if (!impl->pipeline.is_valid())
        return false;

    RDSamplerState sampler_state;
    sampler_state.set_min_filter(RenderingDevice::SAMPLER_FILTER_LINEAR);
    sampler_state.set_mag_filter(RenderingDevice::SAMPLER_FILTER_LINEAR);
    sampler_state.set_mip_filter(RenderingDevice::SAMPLER_FILTER_LINEAR);
    sampler_state.set_repeat_u(RenderingDevice::SAMPLER_REPEAT_MODE_REPEAT);
    sampler_state.set_repeat_v(RenderingDevice::SAMPLER_REPEAT_MODE_REPEAT);
    sampler_state.set_repeat_w(RenderingDevice::SAMPLER_REPEAT_MODE_REPEAT);

    impl->sampler = RD->sampler_create(&sampler_state);

    impl->srcBuffers.resize(3);
    impl->uniforms.resize(1);
    return true;
}

void RdRenderer::Render(RID fb, ImDrawData* drawData)
{
    RenderingDevice* RD = RenderingServer::get_singleton()->get_rendering_device();

    if (!fb.is_valid())
        return;

    godot::PackedFloat32Array pcfloats;
    pcfloats.resize(4);
    pcfloats[0] = 2.0f / drawData->DisplaySize.x;
    pcfloats[1] = 2.0f / drawData->DisplaySize.y;
    pcfloats[2] = -1.0f - (drawData->DisplayPos.x * pcfloats[0]);
    pcfloats[3] = -1.0f - (drawData->DisplayPos.y * pcfloats[1]);
    godot::PackedByteArray pcbuf = pcfloats.to_byte_array();

    // allocate merged index and vertex buffers
    if (impl->idxBufferSize < drawData->TotalIdxCount)
    {
        if (impl->idxBuffer.get_id() != 0)
            RD->free_rid(impl->idxBuffer);
        impl->idxBuffer = RD->index_buffer_create(drawData->TotalIdxCount, RenderingDevice::INDEX_BUFFER_FORMAT_UINT16);
        impl->idxBufferSize = drawData->TotalIdxCount;
    }

    if (impl->vtxBufferSize < drawData->TotalVtxCount)
    {
        if (impl->vtxBuffer.get_id() != 0)
            RD->free_rid(impl->vtxBuffer);
        impl->vtxBuffer = RD->vertex_buffer_create(drawData->TotalVtxCount * sizeof(ImDrawVert));
        impl->vtxBufferSize = drawData->TotalVtxCount;
    }

    // check if our font texture is still valid
    std::erase_if(impl->uniformSets, [RD](const auto& kv) { return !RD->uniform_set_is_valid(kv.second); });

    if (drawData->CmdListsCount > 0)
        impl->SetupBuffers(drawData);

    // draw
    const int64_t dl = RD->draw_list_begin(fb,
                                           RenderingDevice::INITIAL_ACTION_CLEAR,
                                           RenderingDevice::FINAL_ACTION_READ,
                                           RenderingDevice::INITIAL_ACTION_CLEAR,
                                           RenderingDevice::FINAL_ACTION_READ,
                                           impl->clearColors);

    RD->draw_list_bind_render_pipeline(dl, impl->pipeline);
    RD->draw_list_set_push_constant(dl, pcbuf, pcbuf.size());

    int globalIdxOffset = 0;
    int globalVtxOffset = 0;
    for (int i = 0; i < drawData->CmdListsCount; ++i)
    {
        ImDrawList* cmdList = drawData->CmdLists[i];

        for (int cmdi = 0; cmdi < cmdList->CmdBuffer.Size; ++cmdi)
        {
            const ImDrawCmd& drawCmd = cmdList->CmdBuffer[cmdi];
            if (drawCmd.ElemCount == 0)
                continue;
            if (!impl->uniformSets.contains(drawCmd.GetTexID()))
                continue;

            const RID idxArray =
                RD->index_array_create(impl->idxBuffer, drawCmd.IdxOffset + globalIdxOffset, drawCmd.ElemCount);

            const int64_t voff = (drawCmd.VtxOffset + globalVtxOffset) * sizeof(ImDrawVert);
            impl->srcBuffers[0] = impl->srcBuffers[1] = impl->srcBuffers[2] = impl->vtxBuffer;
            impl->vtxOffsets[0] = impl->vtxOffsets[1] = impl->vtxOffsets[2] = voff;
            const RID vtxArray =
                RD->vertex_array_create(cmdList->VtxBuffer.Size, impl->vtxFormat, impl->srcBuffers, impl->vtxOffsets);

            RD->draw_list_bind_uniform_set(dl, impl->uniformSets[drawCmd.GetTexID()], 0);
            RD->draw_list_bind_index_array(dl, idxArray);
            RD->draw_list_bind_vertex_array(dl, vtxArray);

            godot::Rect2 clipRect = {drawCmd.ClipRect.x,
                                     drawCmd.ClipRect.y,
                                     drawCmd.ClipRect.z - drawCmd.ClipRect.x,
                                     drawCmd.ClipRect.w - drawCmd.ClipRect.y};
            clipRect.position -= godot::Vector2i(drawData->DisplayPos.x, drawData->DisplayPos.y);
            RD->draw_list_enable_scissor(dl, clipRect);

            RD->draw_list_draw(dl, true, 1);

            RD->free_rid(idxArray);
            RD->free_rid(vtxArray);
        }
        globalIdxOffset += cmdList->IdxBuffer.Size;
        globalVtxOffset += cmdList->VtxBuffer.Size;
    }
    RD->draw_list_end();
}

RdRenderer::~RdRenderer()
{
    RenderingDevice* RD = RenderingServer::get_singleton()->get_rendering_device();
    RD->free_rid(impl->shader);
    RD->free_rid(impl->sampler);
    if (impl->idxBuffer.is_valid())
        RD->free_rid(impl->idxBuffer);
    if (impl->vtxBuffer.is_valid())
        RD->free_rid(impl->vtxBuffer);
}

void RdRenderer::InitViewport(RID vprid)
{
    RenderingServer::get_singleton()->viewport_set_clear_mode(vprid, RenderingServer::VIEWPORT_CLEAR_NEVER);
}

void RdRenderer::FreeUnusedTextures()
{
    RenderingDevice* RD = RenderingServer::get_singleton()->get_rendering_device();

    // clean up unused textures
    std::vector<ImTextureID> keys;
    keys.reserve(impl->uniformSets.size());
    for (const auto& kv : impl->uniformSets)
        keys.push_back(kv.first);

    for (ImTextureID texid : keys)
    {
        if (!impl->usedTextures.contains(texid))
        {
            RD->free_rid(impl->uniformSets[texid]);
            impl->uniformSets.erase(texid);
        }
    }
    impl->usedTextures.clear();
}

void RdRenderer::Render()
{
    auto& pio = ImGui::GetPlatformIO();
    for (ImGuiViewport* vp : pio.Viewports)
    {
        if (!(vp->Flags & ImGuiViewportFlags_IsMinimized))
        {
            ReplaceTextureRIDs(vp->DrawData);
            const RID vprid = make_rid(vp->RendererUserData);
            Render(GetFramebuffer(vprid), vp->DrawData);
        }
    }
    FreeUnusedTextures();
}

void RdRenderer::ReplaceTextureRIDs(ImDrawData* drawData)
{
    const RenderingServer* RS = RenderingServer::get_singleton();
    for (int i = 0; i < drawData->CmdListsCount; ++i)
    {
        ImDrawList* cmdList = drawData->CmdLists[i];
        for (ImDrawCmd& drawCmd : cmdList->CmdBuffer)
        {
            drawCmd.TextureId = (ImTextureID)RS->texture_get_rd_texture(make_rid(drawCmd.TextureId)).get_id();
        }
    }
}

} // namespace ImGui::Godot

File: addons/imgui-godot/ImGuiGodot/Internal/RdRenderer.cs
using Godot;
using ImGuiNET;
using System;
using System.Buffers;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace ImGuiGodot.Internal;

internal class RdRenderer : IRenderer
{
    protected readonly RenderingDevice RD;
    private readonly Color[] _clearColors = { new Color(0f, 0f, 0f, 0f) };
    private readonly Rid _shader;
    private readonly Rid _pipeline;
    private readonly Rid _sampler;
    private readonly long _vtxFormat;
    private readonly Dictionary<Rid, Rid> _framebuffers = new();
    private readonly float[] _scale = new float[2];
    private readonly float[] _translate = new float[2];
    private readonly byte[] _pcbuf = new byte[16];
    private readonly ArrayPool<byte> _bufPool = ArrayPool<byte>.Create();

    private Rid _idxBuffer;
    private int _idxBufferSize = 0; // size in indices
    private Rid _vtxBuffer;
    private int _vtxBufferSize = 0; // size in vertices

    private readonly Dictionary<IntPtr, Rid> _uniformSets = new(8);
    private readonly HashSet<IntPtr> _usedTextures = new(8);

    private readonly Rect2 _zeroRect = new(new(0f, 0f), new(0f, 0f));
#if GODOT4_1_OR_GREATER
    private readonly Godot.Collections.Array<Rid> _storageTextures = new();
#else
    private readonly Godot.Collections.Array _storageTextures = new();
#endif
    private readonly Godot.Collections.Array<Rid> _srcBuffers = new();
    private readonly long[] _vtxOffsets = new long[3];
    private readonly Godot.Collections.Array<RDUniform> _uniformArray = new();

    public string Name => "imgui_impl_godot4_rd";

    public RdRenderer()
    {
        RD = RenderingServer.GetRenderingDevice();
        if (RD is null)
        {
            throw new PlatformNotSupportedException("failed to get RenderingDevice");
        }

        // set up everything to match the official Vulkan backend as closely as possible

        // compiling from source takes ~400ms, so we use a SPIR-V resource
        using var spirv = ResourceLoader.Load<RDShaderSpirV>("res://addons/imgui-godot/data/ImGuiShaderSPIRV.tres");
        _shader = RD.ShaderCreateFromSpirV(spirv);

#if IMGUI_GODOT_DEV
#pragma warning disable CA2201
        using var src = new RDShaderSource()
        {
            SourceFragment = _fragmentShaderSource,
            SourceVertex = _vertexShaderSource,
        };
        using var freshSpirv = RD.ShaderCompileSpirvFromSource(src);
        if (!System.Linq.Enumerable.SequenceEqual(spirv.BytecodeFragment, freshSpirv.BytecodeFragment))
            throw new Exception("fragment bytecode mismatch");
        if (!System.Linq.Enumerable.SequenceEqual(spirv.BytecodeVertex, freshSpirv.BytecodeVertex))
            throw new Exception("vertex bytecode mismatch");
#pragma warning restore CA2201
#endif

        // create vertex format
        uint vtxStride = (uint)Marshal.SizeOf<ImDrawVert>();

        using RDVertexAttribute attrPoints = new()
        {
            Location = 0,
            Format = RenderingDevice.DataFormat.R32G32Sfloat,
            Stride = vtxStride,
            Offset = 0
        };

        using RDVertexAttribute attrUvs = new()
        {
            Location = 1,
            Format = RenderingDevice.DataFormat.R32G32Sfloat,
            Stride = vtxStride,
            Offset = sizeof(float) * 2
        };

        using RDVertexAttribute attrColors = new()
        {
            Location = 2,
            Format = RenderingDevice.DataFormat.R8G8B8A8Unorm,
            Stride = vtxStride,
            Offset = sizeof(float) * 4
        };

        var vattrs = new Godot.Collections.Array<RDVertexAttribute>() { attrPoints, attrUvs, attrColors };
        _vtxFormat = RD.VertexFormatCreate(vattrs);

        // blend state
        using var bsa = new RDPipelineColorBlendStateAttachment
        {
            EnableBlend = true,

            SrcColorBlendFactor = RenderingDevice.BlendFactor.SrcAlpha,
            DstColorBlendFactor = RenderingDevice.BlendFactor.OneMinusSrcAlpha,
            ColorBlendOp = RenderingDevice.BlendOperation.Add,

            SrcAlphaBlendFactor = RenderingDevice.BlendFactor.One,
            DstAlphaBlendFactor = RenderingDevice.BlendFactor.OneMinusSrcAlpha,
            AlphaBlendOp = RenderingDevice.BlendOperation.Add,
        };

        using var blendData = new RDPipelineColorBlendState
        {
            BlendConstant = new Color(0, 0, 0, 0),
        };
        blendData.Attachments.Add(bsa);

        // rasterization state
        using var rasterizationState = new RDPipelineRasterizationState
        {
            FrontFace = RenderingDevice.PolygonFrontFace.CounterClockwise
        };

        // pipeline
        _pipeline = RD.RenderPipelineCreate(
            _shader,
            RD.ScreenGetFramebufferFormat(),
            _vtxFormat,
            RenderingDevice.RenderPrimitive.Triangles,
            rasterizationState,
            new RDPipelineMultisampleState(),
            new RDPipelineDepthStencilState(),
            blendData);

        // sampler used for all textures
        using var samplerState = new RDSamplerState
        {
            MinFilter = RenderingDevice.SamplerFilter.Linear,
            MagFilter = RenderingDevice.SamplerFilter.Linear,
            MipFilter = RenderingDevice.SamplerFilter.Linear,
            RepeatU = RenderingDevice.SamplerRepeatMode.Repeat,
            RepeatV = RenderingDevice.SamplerRepeatMode.Repeat,
            RepeatW = RenderingDevice.SamplerRepeatMode.Repeat
        };
        _sampler = RD.SamplerCreate(samplerState);

        _srcBuffers.Resize(3);
        _uniformArray.Resize(1);
    }

    public void Init(ImGuiIOPtr io)
    {
        io.BackendFlags |= ImGuiBackendFlags.RendererHasVtxOffset;
        io.BackendFlags |= ImGuiBackendFlags.RendererHasViewports;
    }

    public void InitViewport(Rid vprid)
    {
        //RenderingServer.ViewportSetUpdateMode(vprid, RenderingServer.ViewportUpdateMode.Disabled);
        RenderingServer.ViewportSetClearMode(vprid, RenderingServer.ViewportClearMode.Never);
    }

    public void CloseViewport(Rid vprid)
    {
    }

    private void SetupBuffers(ImDrawDataPtr drawData)
    {
        int vertSize = Marshal.SizeOf<ImDrawVert>();
        int globalIdxOffset = 0;
        int globalVtxOffset = 0;

        // set up buffers
        int idxBufSize = drawData.TotalIdxCount * sizeof(ushort);
        byte[] idxBuf = _bufPool.Rent(idxBufSize);
        int vertBufSize = drawData.TotalVtxCount * vertSize;
        byte[] vertBuf = _bufPool.Rent(vertBufSize);
        for (int i = 0; i < drawData.CmdLists.Size; ++i)
        {
            ImDrawListPtr cmdList = drawData.CmdLists[i];

            int vertBytes = cmdList.VtxBuffer.Size * vertSize;
            Marshal.Copy(cmdList.VtxBuffer.Data, vertBuf, globalVtxOffset, vertBytes);
            globalVtxOffset += vertBytes;

            int idxBytes = cmdList.IdxBuffer.Size * sizeof(ushort);
            Marshal.Copy(cmdList.IdxBuffer.Data, idxBuf, globalIdxOffset, idxBytes);
            globalIdxOffset += idxBytes;

            // create a uniform set for each texture
            for (int cmdi = 0; cmdi < cmdList.CmdBuffer.Size; ++cmdi)
            {
                ImDrawCmdPtr drawCmd = cmdList.CmdBuffer[cmdi];
                IntPtr texid = drawCmd.GetTexID();
                if (texid == IntPtr.Zero)
                    continue;
                Rid texrid = Util.ConstructRid((ulong)texid);
                if (!RD.TextureIsValid(texrid))
                    continue;

                _usedTextures.Add(texid);
                if (!_uniformSets.ContainsKey(texid))
                {
                    using RDUniform uniform = new()
                    {
                        Binding = 0,
                        UniformType = RenderingDevice.UniformType.SamplerWithTexture
                    };
                    uniform.AddId(_sampler);
                    uniform.AddId(texrid);
                    _uniformArray[0] = uniform;
                    Rid uniformSet = RD.UniformSetCreate(_uniformArray, _shader, 0);
                    _uniformSets[texid] = uniformSet;
                }
            }
        }
        RD.BufferUpdate(_idxBuffer, 0, (uint)idxBufSize, idxBuf);
        _bufPool.Return(idxBuf);
        RD.BufferUpdate(_vtxBuffer, 0, (uint)vertBufSize, vertBuf);
        _bufPool.Return(vertBuf);
    }

    protected static void ReplaceTextureRids(ImDrawDataPtr drawData)
    {
        for (int i = 0; i < drawData.CmdLists.Size; ++i)
        {
            ImDrawListPtr cmdList = drawData.CmdLists[i];
            for (int cmdi = 0; cmdi < cmdList.CmdBuffer.Size; ++cmdi)
            {
                ImDrawCmdPtr drawCmd = cmdList.CmdBuffer[cmdi];
                drawCmd.TextureId = (IntPtr)RenderingServer.TextureGetRdTexture(Util.ConstructRid((ulong)drawCmd.TextureId)).Id;
            }
        }
    }

    protected void FreeUnusedTextures()
    {
        // clean up unused textures
        foreach (IntPtr texid in _uniformSets.Keys)
        {
            if (!_usedTextures.Contains(texid))
            {
                RD.FreeRid(_uniformSets[texid]);
                _uniformSets.Remove(texid);
            }
        }
        _usedTextures.Clear();
    }

    public void RenderDrawData()
    {
        var pio = ImGui.GetPlatformIO();
        for (int i = 0; i < pio.Viewports.Size; ++i)
        {
            var vp = pio.Viewports[i];
            if (!vp.Flags.HasFlag(ImGuiViewportFlags.IsMinimized))
            {
                ReplaceTextureRids(vp.DrawData);
                Rid vprid = Util.ConstructRid((ulong)vp.RendererUserData);
                RenderOne(GetFramebuffer(vprid), vp.DrawData);
            }
        }
        FreeUnusedTextures();
    }

    protected void RenderOne(Rid fb, ImDrawDataPtr drawData)
    {
#if IMGUI_GODOT_DEV
        RD.DrawCommandBeginLabel("ImGui", Colors.Purple);
#endif

        if (!fb.IsValid)
            return;

        int vertSize = Marshal.SizeOf<ImDrawVert>();

        _scale[0] = 2.0f / drawData.DisplaySize.X;
        _scale[1] = 2.0f / drawData.DisplaySize.Y;

        _translate[0] = -1.0f - (drawData.DisplayPos.X * _scale[0]);
        _translate[1] = -1.0f - (drawData.DisplayPos.Y * _scale[1]);

        Buffer.BlockCopy(_scale, 0, _pcbuf, 0, 8);
        Buffer.BlockCopy(_translate, 0, _pcbuf, 8, 8);

        // allocate merged index and vertex buffers
        if (_idxBufferSize < drawData.TotalIdxCount)
        {
            if (_idxBuffer.Id != 0)
                RD.FreeRid(_idxBuffer);
            _idxBuffer = RD.IndexBufferCreate((uint)drawData.TotalIdxCount, RenderingDevice.IndexBufferFormat.Uint16);
            _idxBufferSize = drawData.TotalIdxCount;
        }

        if (_vtxBufferSize < drawData.TotalVtxCount)
        {
            if (_vtxBuffer.Id != 0)
                RD.FreeRid(_vtxBuffer);
            _vtxBuffer = RD.VertexBufferCreate((uint)(drawData.TotalVtxCount * vertSize));
            _vtxBufferSize = drawData.TotalVtxCount;
        }

        // check if our font texture is still valid
        foreach (var kv in _uniformSets)
        {
            if (!RD.UniformSetIsValid(kv.Value))
                _uniformSets.Remove(kv.Key);
        }

        if (drawData.CmdListsCount > 0)
            SetupBuffers(drawData);

        // draw
        long dl = RD.DrawListBegin(fb,
                RenderingDevice.InitialAction.Clear, RenderingDevice.FinalAction.Read,
                RenderingDevice.InitialAction.Clear, RenderingDevice.FinalAction.Read,
                _clearColors, 1f, 0, _zeroRect, _storageTextures);

        RD.DrawListBindRenderPipeline(dl, _pipeline);
        RD.DrawListSetPushConstant(dl, _pcbuf, (uint)_pcbuf.Length);

        int globalIdxOffset = 0;
        int globalVtxOffset = 0;
        for (int i = 0; i < drawData.CmdLists.Size; ++i)
        {
            ImDrawListPtr cmdList = drawData.CmdLists[i];

            for (int cmdi = 0; cmdi < cmdList.CmdBuffer.Size; ++cmdi)
            {
                ImDrawCmdPtr drawCmd = cmdList.CmdBuffer[cmdi];
                if (drawCmd.ElemCount == 0)
                    continue;
                if (!_uniformSets.ContainsKey(drawCmd.GetTexID()))
                    continue;

                Rid idxArray = RD.IndexArrayCreate(_idxBuffer,
                    (uint)(drawCmd.IdxOffset + globalIdxOffset),
                    drawCmd.ElemCount);

                long voff = (drawCmd.VtxOffset + globalVtxOffset) * vertSize;
                _srcBuffers[0] = _srcBuffers[1] = _srcBuffers[2] = _vtxBuffer;
                _vtxOffsets[0] = _vtxOffsets[1] = _vtxOffsets[2] = voff;
                Rid vtxArray = RD.VertexArrayCreate(
                    (uint)cmdList.VtxBuffer.Size,
                    _vtxFormat,
                    _srcBuffers,
                    _vtxOffsets);

                RD.DrawListBindUniformSet(dl, _uniformSets[drawCmd.GetTexID()], 0);
                RD.DrawListBindIndexArray(dl, idxArray);
                RD.DrawListBindVertexArray(dl, vtxArray);

                var clipRect = new Rect2(
                    drawCmd.ClipRect.X,
                    drawCmd.ClipRect.Y,
                    drawCmd.ClipRect.Z - drawCmd.ClipRect.X,
                    drawCmd.ClipRect.W - drawCmd.ClipRect.Y);
                clipRect.Position -= drawData.DisplayPos.ToVector2I();
                RD.DrawListEnableScissor(dl, clipRect);

                RD.DrawListDraw(dl, true, 1);

                RD.FreeRid(idxArray);
                RD.FreeRid(vtxArray);
            }
            globalIdxOffset += cmdList.IdxBuffer.Size;
            globalVtxOffset += cmdList.VtxBuffer.Size;
        }
        RD.DrawListEnd();
#if IMGUI_GODOT_DEV
        RD.DrawCommandEndLabel();
#endif
    }

    public void OnHide()
    {
    }

    public void Shutdown()
    {
        RD.FreeRid(_sampler);
        RD.FreeRid(_shader);
        if (_idxBuffer.Id != 0)
            RD.FreeRid(_idxBuffer);
        if (_vtxBuffer.Id != 0)
            RD.FreeRid(_vtxBuffer);
    }

    protected Rid GetFramebuffer(Rid vprid)
    {
        if (!vprid.IsValid)
            return new Rid();

        if (_framebuffers.TryGetValue(vprid, out Rid fb))
        {
            if (RD.FramebufferIsValid(fb))
                return fb;
        }

        Rid vptex = RenderingServer.TextureGetRdTexture(RenderingServer.ViewportGetTexture(vprid));
        fb = RD.FramebufferCreate(new() { vptex });
        _framebuffers[vprid] = fb;
        return fb;
    }

#if IMGUI_GODOT_DEV
    // shader source borrowed from imgui_impl_vulkan.cpp
    private static readonly string _vertexShaderSource = @"#version 450 core
layout(location = 0) in vec2 aPos;
layout(location = 1) in vec2 aUV;
layout(location = 2) in vec4 aColor;
layout(push_constant) uniform uPushConstant { vec2 uScale; vec2 uTranslate; } pc;

out gl_PerVertex { vec4 gl_Position; };
layout(location = 0) out struct { vec4 Color; vec2 UV; } Out;

void main()
{
    Out.Color = aColor;
    Out.UV = aUV;
    gl_Position = vec4(aPos * pc.uScale + pc.uTranslate, 0, 1);
}";

    private static readonly string _fragmentShaderSource = @"#version 450 core
layout(location = 0) out vec4 fColor;
layout(set=0, binding=0) uniform sampler2D sTexture;
layout(location = 0) in struct { vec4 Color; vec2 UV; } In;
void main()
{
    fColor = In.Color * texture(sTexture, In.UV.st);
}";
#endif
}
